Remotion/Core/Core/Context/AsyncLocalStorageProvider.cs

```diff
@@ -14,38 +14,199 @@
 // You should have received a copy of the GNU Lesser General Public License
 // along with re-motion; if not, see http://www.gnu.org/licenses.
 //
+
+// To help with debugging uncomment the next line to get output whenever the AsyncLocalValueChangedHandler is called
+// #define TRACE_ASYNCLOCAL
+
 using System;
 using System.Collections;
+using System.Diagnostics;
 using System.Threading;
 using Remotion.ServiceLocation;
 using Remotion.Utilities;
 
+#if TRACE_ASYNCLOCAL
+using System.Runtime.CompilerServices;
+#endif
+
 namespace Remotion.Context
 {
   [ImplementationFor (typeof (ISafeContextStorageProvider), Position = 1)]
```

Comment:
We don't have instance state in either implementation of ISafeContextStorageProvider -> please verify and then add Lifetime = Singleton to the attribute. 

By doing this, we 1. don't change the semantics, 2. improve performance (side benefit), and 3. enable us to register a special instance that supports the Trace_AsyncLocal enabled via boolean flag instead of compiler.

So, next step: 
Add a private ctor that allows setting a boolean flag to enable the Trace-semantics. An application developer is then able to simply reconfigure the system to enable tracing.

Reply:
How is the application developer supposed to enable tracing if the constructor is private?

Reply:
Forgot: and add a factory method with a good name.

Reply:
So public constructor that calls private constructor with `false`, Private constructor taking tracing boolean, and a factory method that calls the private constructor with tracing set to `true`?